sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4

```diff
@@ -720,14 +725,18 @@ qualifiedName
 
 identifier
     : strictIdentifier
+    // We can't move the tokens below into `strictIdentifier` because we need to parse some SQL texts,
+    // e.g., SELECT * FROM t1 CROSS JOIN t2
     | ANTI | FULL | INNER | LEFT | SEMI | RIGHT | NATURAL | JOIN | CROSS | ON
     | UNION | INTERSECT | EXCEPT | SETMINUS
     ;
 
 strictIdentifier
-    : IDENTIFIER             #unquotedIdentifier
-    | quotedIdentifier       #quotedIdentifierAlternative
-    | nonReserved            #unquotedIdentifier
+    : IDENTIFIER              #unquotedIdentifier
+    | quotedIdentifier        #quotedIdentifierAlternative
+    | {ansi}? ansiReserved    #unquotedIdentifier
```

Comment:
how can reserved keywords appear in `strictIdentifier`?

Reply:
now I'm totally confused. What I expect to see is
```
identifier
    : strictIdentifier
    | {ansi}? ansiReserved
    | {!ansi}? reserved

strictIdentifier:
    : xxxx
    | {ansi}? ansiNonReserved
    | {!ansi}? nonReserved
```

Then it's very clear that how we distinguish reserved and non-reserved keywords and how do they change with the "ansi" config.

Reply:
You describe the case where `ansi` = false?

Reply:
when ansi=false, it should be the same as before
```
identifier
    : strictIdentifier
    | reserved

strictIdentifier:
    : xxxx
    | nonReserved
```

This is literally same as previous, we just do a code refactor put move something out and name it "reserved".

Reply:
```
identifier
    : strictIdentifier
    | {ansi}? ansiReserved
    | {!ansi}? reserved

strictIdentifier:
    : xxxx
    | {ansi}? ansiNonReserved
    | {!ansi}? nonReserved
```
In this rule set with ansi=true, `sql("select 1 as select from (select 1)")` throws an exception cuz of the reserved keyword `select`, but `sql("select 1 from (select 1) as select")` does not. So, I moved `ansi}? ansiReserved` into `strictIdentifier`....

Reply:
> `sql("select 1 as select from (select 1)")` throws an exception cuz of the reserved keyword `select`

Isn't it what we expect? That's how reserved keywords should behave, right?

Reply:
maybe in the .md doc we should clear mention what's the difference between reserved and non-reserved keywords.

Reply:
Aha, I see. If so, looks ok to me. I just follow the PostgreSQL behaivour;
```
postgres=# select * from t select;
ERROR:  syntax error at or near "select" at character 17
STATEMENT:  select * from t select;
ERROR:  syntax error at or near "select"
LINE 1: select * from t select;
```
In the pg, we cannot use reserved keywords in a table alias.

Reply:
> maybe in the .md doc we should clear mention what's the difference between reserved and non-reserved keywords.

ok, I'll brush up the doc.

Reply:
cool thanks! I think this is very close!

Reply:
ok! Fixed.